### ISSUE TITLE
when veritesting is true, I  get more found paths than I set

### ISSUE THREAD

**Issue by brokendragon** [angr/angr#169]
Hi,
when I set ```veritesting=True ``` in path_group and set ```num_find=1 ```  in path_group .explore, I get more than one ```found``` results. There is any parameter that I can set to just get one found path when veritesting is True?
Below is the commands:
 ```
    p=b.factory.path(args=[arg0,arg1,arg2,arg3])
    pg=b.factory.path_group(p,veritesting=True)
    pg=pg.explore(find='0x4006f5',num_find=1)
```

here is the result:
 ```
<PathGroup with 1 active, 2 found>
```
Thanks a lot!

**Comment by ltfish** [comment]
Can you please tell me if those two found paths are the same (with the exact same trace)?

I've never noticed this issue before (which is really because I never paid any attention to `num_find` while using Veritesting). My guess is, that with Veritesting enabled, it does respect the `num_find` setting during its core static symbolic execution procedure (which is implemented in `angr/analyses/veritesting.py`). When it decides to exit from static execution phase, `PathGroup` checks if there are more than `num_find` paths found, and if there are, it quits from the iteration and returns all the result. And it's totally possible to get more than one `found` paths during static symbolic execution.

If my guess is right, the solution would be passing `num_find` into Veritesting.

**Comment by brokendragon** [comment]
The last two blocks in those two found path are defferent.
below is my tested program.
``` 
void test_swit(){}
int main(int argc, char** argv)
{
  int y=argc;
  if(y<0x500)
  { 
	  test_swit();
  }
  else if(y>0x800){	
	  test_swit();
  }
  return 0; 
}
```
And I changed a little in the commands
```
    sargc = claripy.BVS('sargc', 64)
    p=b.factory.path(argc=sargc)
    pg=b.factory.path_group(p,veritesting=True)
    pg_1=pg.explore(find=ad,num_find=1)
```
below is the trace of two found path.
```
<PathGroup with 1 active, 2 found>

pt_1: ('<SimIRSB 0x400400>', '<SimIRSB 0x4003e0>', '<SimProcedure __libc_start_main>', '<SimIRSB 0x400550>', '<SimIRSB 0x4003a8>', '<SimIRSB 0x4003bd>', '<SimIRSB 0x400583>', '<SimIRSB 0x400588>', '<SimIRSB 0x4004d0>', '<SimIRSB 0x4004db>', '<SimIRSB 0x400470>', '<SimIRSB 0x4004a8>', '<SimIRSB 0x40059d>', '<SimIRSB 0x4005a6>', '<SimProcedure __libc_start_main>', '<SimIRSB 0x4004fd>', '<SimIRSB 0x40051b>')

pt_2: ('<SimIRSB 0x400400>', '<SimIRSB 0x4003e0>', '<SimProcedure __libc_start_main>', '<SimIRSB 0x400550>', '<SimIRSB 0x4003a8>', '<SimIRSB 0x4003bd>', '<SimIRSB 0x400583>', '<SimIRSB 0x400588>', '<SimIRSB 0x4004d0>', '<SimIRSB 0x4004db>', '<SimIRSB 0x400470>', '<SimIRSB 0x4004a8>', '<SimIRSB 0x40059d>', '<SimIRSB 0x4005a6>', '<SimProcedure __libc_start_main>', '<SimIRSB 0x4004fd>', '<SimIRSB 0x400527>', '<SimIRSB 0x400530>')
```
Thanks for response!

**Comment by zardus** [comment]
This looks like an issue that we might have fixed recently with a rewrite of our Veritesting support. We'll be releasing that in a little over a week; stay tuned!

**Comment by rhelmot** [comment]
So the bottom line is that this has nothing to do with veritesting. `num_find` is not a guarantee that the exploration will finish with exactly that many found paths. Hell, it's not even a guarantee that the exploration will finish with _at least_ that many found paths. The only thing that you get by setting `num_find` is asking the exploration to stop as soon as it sees it's found at least that many new paths in the `found` stash.

Here's an example without veritesting where setting `num_find=1` will get you two found paths:

test2.c:
```c
#include <stdio.h>

int main(int argc, char **argv) {
  if (argc == 1) return 1;
  else if (argc == 2) puts("bbb");
  else puts("ccc");
  return 0;
}
```

test2.py:
```python
import claripy, angr
p = angr.Project('test2')
pg = p.factory.path_group(p.factory.entry_state(argc=claripy.BVS('argc', 64)))
pg.explore(find=0x400400, num_find=1)  # address of puts
print pg
```

output: `<PathGroup with 1 active, 2 found>`

What's happening is that the "is this path found" check happens after all the paths in the active stash have been stepped forward. So, if in a single analysis step, more than one path ends up satisfying the `find` condition, then both of those paths will end up in the found stash. Then, between analysis steps, we check if the termination condition is satisfied - in this case the termination condition is `len(pg.found) >= 1`, so we stop.

Basically, fixing this would require a serious reorganization of the internal path group code for not a whole lot of gain. I'm going to close this, but if you've got a good reason you need `explore` to always give you exactly the number of paths you asked for, then we'll see about getting you a custom solution.
